File: CPP/ProfitableSchemes.cpp
#include <algorithm>
#include <iostream>
#include <vector>

class Solution {
public:
    int profitableSchemes(int n, int minProfit, std::vector<int>& group, std::vector<int>& profit) {
        constexpr int MOD = 1'000'000'007;
        std::vector<std::vector<int>> dp(n + 1, std::vector<int>(minProfit + 1, 0));
        dp[0][0] = 1;
        for (size_t i = 0; i < group.size(); ++i) {
            int g = group[i], p = profit[i];
            for (int j = n; j >= g; --j) {
                for (int k = minProfit; k >= 0; --k) {
                    if (!dp[j - g][k]) continue;
                    int nk = std::min(k + p, minProfit);
                    dp[j][nk] = (dp[j][nk] + dp[j - g][k]) % MOD;
                }
            }
        }
        long long total = 0;
        for (int j = 0; j <= n; ++j) total = (total + dp[j][minProfit]) % MOD;
        return (int)total;
    }
};

int main() {
    Solution s;
    std::vector<int> g1 = {2, 2}, p1 = {2, 3};
    std::vector<int> g2 = {2, 3, 5}, p2 = {6, 7, 8};
    std::vector<int> g3 = {1, 1}, p3 = {1, 1};
    std::cout << s.profitableSchemes(5, 3, g1, p1) << "\n";  // 2
    std::cout << s.profitableSchemes(10, 5, g2, p2) << "\n"; // 7
    std::cout << s.profitableSchemes(1, 1, g3, p3) << "\n";  // 2
    return 0;
}

File: Java/ProfitableSchemes.java
public class ProfitableSchemes {
    private static final int MOD = 1_000_000_007;

    public int profitableSchemes(int n, int minProfit, int[] group, int[] profit) {
        int[][] dp = new int[n + 1][minProfit + 1];
        dp[0][0] = 1;
        for (int i = 0; i < group.length; i++) {
            int g = group[i], p = profit[i];
            for (int j = n; j >= g; j--) {
                for (int k = minProfit; k >= 0; k--) {
                    if (dp[j - g][k] == 0) continue;
                    int nk = Math.min(k + p, minProfit);
                    dp[j][nk] = (dp[j][nk] + dp[j - g][k]) % MOD;
                }
            }
        }
        long total = 0;
        for (int j = 0; j <= n; j++) total = (total + dp[j][minProfit]) % MOD;
        return (int) total;
    }

    public static void main(String[] args) {
        ProfitableSchemes s = new ProfitableSchemes();
        System.out.println(s.profitableSchemes(5, 3, new int[] { 2, 2 }, new int[] { 2, 3 }));        // 2
        System.out.println(s.profitableSchemes(10, 5, new int[] { 2, 3, 5 }, new int[] { 6, 7, 8 })); // 7
        System.out.println(s.profitableSchemes(1, 1, new int[] { 1, 1 }, new int[] { 1, 1 }));        // 2
    }
}
